### PR TITLE
refactor(components): [carousel] refactor

### PR DESCRIPTION
**Pull request by buqiyuan**
Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

File: packages/components/carousel/src/carousel.vue [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+} from 'vue'
+import { throttle } from 'lodash-unified'
+import { useResizeObserver } from '@vueuse/core'
+import { debugWarn } from '@element-plus/utils'
+import { ElIcon } from '@element-plus/components/icon'
+import { ArrowLeft, ArrowRight } from '@element-plus/icons-vue'
+import { useNamespace } from '@element-plus/hooks'
+import { carouselContextKey } from '@element-plus/tokens'
+import { carouselProps, carouselEmits } from './carousel'
+import type { CarouselItemContext } from '@element-plus/tokens'
```
Comment (JeremyWuuuuu) [inline packages/components/carousel/src/carousel.vue:82]: **Nit** insert an blank line before line#82

File: packages/components/carousel/src/carousel.vue [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+const playSlides = () => {
+  if (activeIndex.value < items.value.length - 1) {
+    activeIndex.value = activeIndex.value + 1
+  } else if (props.loop) {
+    activeIndex.value = 0
+  }
+}
+
+function setActiveItem(index) {
+  if (typeof index === 'string') {
```
Comment (JeremyWuuuuu) [inline packages/components/carousel/src/carousel.vue:160]: **Nit**

```suggestion
  if (isString(index)) {
```

File: packages/components/carousel/src/carousel.vue [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+  })
+})
+
+// provide
+provide(carouselContextKey, {
+  root,
+  direction: props.direction,
+  type: props.type,
+  items,
+  loop: props.loop,
```
Comment (JeremyWuuuuu) [inline packages/components/carousel/src/carousel.vue:310]: These action removes the reactivity of `direction`, `type`, `loop`. We can do something like:


```suggestion
  direction: toRef(props, 'direction'),
  type: toRef(props, 'type'),
  items,
  loop: toRef(props, 'loop'),
```

**Review by jw-foss — COMMENTED** [review]
@buqiyuan In general this looks good, I just left a few comments for you to consider. 🏓

**Review by jw-foss — APPROVED** [review]

**Comment by JeremyWuuuuu** [comment]
I am going to approve this PR and address the comments in another one myself. Since this block my work in #6657

**Pull request merged**